src/EditorFeatures/Core/Implementation/CommentSelection/AbstractCommentSelectionCommandHandler.cs

```diff
@@ -0,0 +1,156 @@
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Operations;
+using Roslyn.Utilities;
+using VSCommanding = Microsoft.VisualStudio.Commanding;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.CommentSelection
+{
+    internal enum Operation { Comment, Uncomment, Toggle }
+
+    abstract class AbstractCommentSelectionCommandHandler
+    {
+        protected readonly ITextUndoHistoryRegistry _undoHistoryRegistry;
+        protected readonly IEditorOperationsFactoryService _editorOperationsFactoryService;
+
+        internal AbstractCommentSelectionCommandHandler(
+            ITextUndoHistoryRegistry undoHistoryRegistry,
+            IEditorOperationsFactoryService editorOperationsFactoryService)
+        {
+            Contract.ThrowIfNull(undoHistoryRegistry);
+            Contract.ThrowIfNull(editorOperationsFactoryService);
+
+            _undoHistoryRegistry = undoHistoryRegistry;
+            _editorOperationsFactoryService = editorOperationsFactoryService;
+        }
+
+        protected static VSCommanding.CommandState GetCommandState(ITextBuffer buffer)
+        {
+            if (!buffer.CanApplyChangeDocumentToWorkspace())
+            {
+                return VSCommanding.CommandState.Unspecified;
+            }
+
+            return VSCommanding.CommandState.Available;
+        }
+
+        protected static void Format(ICommentSelectionService service, ITextSnapshot snapshot, IEnumerable<CommentTrackingSpan> changes, CancellationToken cancellationToken)
+        {
+            var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return;
+            }
+
+            // Only format uncomment actions.
+            var textSpans = changes
+                .Where(change => change.Operation == Operation.Uncomment)
+                .Select(uncommentChange => uncommentChange.ToSnapshotSpan(snapshot).Span.ToTextSpan())
+                .ToImmutableArray();
+            var newDocument = service.FormatAsync(document, textSpans, cancellationToken).WaitAndGetResult(cancellationToken);
+            newDocument.Project.Solution.Workspace.ApplyDocumentChanges(newDocument, cancellationToken);
+        }
+
+        protected static void InsertText(List<TextChange> textChanges, int position, string text)
+        {
+            textChanges.Add(new TextChange(new TextSpan(position, 0), text));
+        }
+
+        protected static void DeleteText(List<TextChange> textChanges, TextSpan span)
+        {
+            textChanges.Add(new TextChange(span, string.Empty));
+        }
+
+        internal bool ExecuteCommand(ITextView textView, ITextBuffer subjectBuffer, Operation operation, CommandExecutionContext context)
+        {
+            var title = GetTitle(operation);
+            var message = GetMessage(operation);
+
+            using (context.OperationContext.AddScope(allowCancellation: false, message))
```

Comment:
why is cancellation not allowed? What if this operation takes a very long time?

Reply:
Left this as it was in the comment uncomment handler - I'd rather not get into a situation where a selection has comments partially added or removed.

Reply:
that would not happen.  cancellation would affect computation of the work.  We wouldn't cancel midway through application of the result.